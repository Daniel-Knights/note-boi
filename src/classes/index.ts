export * from './appError';
export * from './debounceQueue';
export * from './encryptor';
export * from './fetchBuilder';
export * from './keyStore';
export * from './storage';
export * from './unsyncedNotesManager';
