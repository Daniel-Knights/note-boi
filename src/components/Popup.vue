<template>
  <Teleport to="#app">
    <div id="popup">
      <div class="popup__content">
        <slot />
      </div>
    </div>
  </Teleport>
</template>

<script lang="ts" setup>
import { onBeforeUnmount } from 'vue';

const emit = defineEmits(['close']);

function closePopup() {
  document.removeEventListener('keydown', keyboardCloseHandler);
  document.removeEventListener('mousedown', clickCloseHandler);

  emit('close');
}

function keyboardCloseHandler(event: KeyboardEvent) {
  if (event.key === 'Escape') closePopup();
}
function clickCloseHandler(event: MouseEvent) {
  if (!(event.target as HTMLElement)?.closest('.popup__content')) {
    closePopup();
  }
}

// setTimeout prevents immediate close
setTimeout(() => {
  document.addEventListener('keydown', keyboardCloseHandler);
  document.body.addEventListener('mousedown', clickCloseHandler);
});

onBeforeUnmount(closePopup);
</script>

<style lang="scss" scoped>
#popup {
  @include v.flex-x(center, center);
  position: fixed;
  @include v.cover;
  @include v.equal-dimensions(100%);
  background-color: rgba(0, 0, 0, 0.5);
  box-shadow: 0 0 10px rgba(0, 0, 0, 0.5);
  z-index: 100;
}

.popup__content {
  padding: 1em;
  max-width: 400px;
  color: var(--colour__primary);
  background-color: var(--colour__secondary);
}
</style>
