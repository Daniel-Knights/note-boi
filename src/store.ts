import { reactive } from 'vue';
import { invoke } from '@tauri-apps/api/tauri';
import { v4 as uuidv4 } from 'uuid';

import { isEmptyNote } from './utils';

export class Note {
  readonly id = uuidv4();
  body = {
    delta: '',
    text: '',
  };
  timestamp = Date.now();
}

interface State {
  notes: Note[];
  selectedNote: Note;
}

export const state = reactive<State>({
  notes: [],
  selectedNote: new Note(),
});

/** Sorts notes in descending order by timestamp. */
function sortNotes() {
  state.notes.sort((a, b) => b.timestamp - a.timestamp);
}

/** Finds note index within {@link state.notes}. */
function findNoteIndex(id: string) {
  return state.notes.findIndex((nt) => nt.id === id);
}

/** Finds note within {@link state.notes}. */
export function findNote(id: string): Note | undefined {
  return state.notes.find((nt) => nt.id === id);
}

/** Deletes {@link state.selectedNote} when both `title` and `body` are empty. */
function clearEmptyNote(): void {
  const isValidClear = state.notes.length > 1;

  const foundNote = findNote(state.selectedNote.id);
  if (!foundNote) return;

  if (isValidClear && isEmptyNote(foundNote)) {
    deleteNote(state.selectedNote.id);
  }
}

const selectEvent = new CustomEvent('note-select');

/**
 * Looks for note with given `id` in {@link state.notes},
 * and sets it to {@link state.selectedNote}.
 */
export function selectNote(id: string): void {
  if (state.selectedNote.id === id) return;

  clearEmptyNote();

  const foundNote = findNote(id);
  if (foundNote) state.selectedNote = { ...foundNote };

  document.dispatchEvent(selectEvent);
}

/** Fetches all notes and updates {@link state}. */
export async function getAllNotes(): Promise<void> {
  const fetchedNotes = await invoke<Note[]>('get_all_notes').catch(console.error);

  const hasNotes = fetchedNotes && fetchedNotes.length > 0;
  if (!hasNotes) return newNote();

  state.notes = fetchedNotes;

  sortNotes();

  if (isEmptyNote(fetchedNotes[0])) {
    state.notes[0].timestamp = Date.now();
  }

  state.selectedNote = { ...state.notes[0] };

  document.dispatchEvent(selectEvent);
}

/** Deletes note with the given `id`. */
export function deleteNote(id: string): void {
  state.notes.splice(findNoteIndex(id), 1);
  state.selectedNote = { ...state.notes[0] };

  sortNotes();

  invoke('delete_note', { id }).catch(console.error);
}

/** Creates an empty note. */
export function newNote(): void {
  const foundNote = findNote(state.selectedNote.id);

  // Only update timestamp if selected note is empty
  if (foundNote && isEmptyNote(foundNote)) {
    state.selectedNote.timestamp = Date.now();
    return;
  }

  const freshNote = new Note();

  state.notes.unshift(freshNote);
  state.selectedNote = { ...freshNote };

  invoke('new_note', { note: { ...freshNote } }).catch(console.error);

  document.dispatchEvent(selectEvent);
}

/** Edits note body on Quill `text-change`. */
export function editBody(delta: string, text: string): void {
  const foundNote = findNote(state.selectedNote.id);
  if (!foundNote || text === foundNote.body.text) return;

  const timestamp = Date.now();
  foundNote.timestamp = timestamp;
  state.selectedNote.timestamp = timestamp;

  foundNote.body = { delta, text };

  sortNotes();

  invoke('edit_note', { note: { ...foundNote } }).catch(console.error);
}
