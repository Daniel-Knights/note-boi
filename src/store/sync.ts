import { reactive } from 'vue';
import { http, invoke } from '@tauri-apps/api';
import type { Response } from '@tauri-apps/api/http';

import { getAllNotes, Note, state as noteState } from './note';
import { isDev, tauriEmit } from '../utils';

export enum ErrorType {
  None,
  Auth,
  Push,
  Pull,
  Logout,
}

interface State {
  username: string;
  password: string;
  token: string;
  hasUnsyncedNotes: boolean;
  isLoading: boolean;
  isLogin: boolean; // For switching login/signup form
  autoSyncEnabled: boolean;
  error: {
    type: ErrorType;
    message: string;
  };
}

const autoSync = localStorage.getItem('auto-sync') as 'true' | 'false' | null;

export const state = reactive<State>({
  username: localStorage.getItem('username') || '',
  password: '',
  token: localStorage.getItem('token') || '',
  hasUnsyncedNotes: false,
  isLoading: false,
  isLogin: true,
  autoSyncEnabled: autoSync !== null ? autoSync === 'true' : true,
  error: {
    type: ErrorType.None,
    message: '',
  },
});

// Utils //

/** Parses an error response and returns a formatted message. */
function parseErrorRes(res: Response<Record<string, unknown>>) {
  const unknownErrorMessage = 'Unknown error, please try again';

  if (!res.data) return unknownErrorMessage;

  return typeof res.data.error === 'string' ? res.data.error : unknownErrorMessage;
}

/** Resets {@link state.error}. */
export function resetError(): void {
  state.error = { type: ErrorType.None, message: '' };
}

/** Wrapper for {@link http.fetch}. */
function tauriFetch<T>(
  endpoint: string,
  method: 'POST' | 'PUT',
  payload?: Record<string, unknown>
): Promise<Response<T>> {
  const baseUrl = isDev() ? 'http://localhost:8000' : 'TODO';

  return http.fetch<T>(`${baseUrl}/api${endpoint}`, {
    method,
    body: {
      type: 'Json',
      payload,
    },
  });
}

function clientSideLogout() {
  state.token = '';
  state.username = '';

  localStorage.removeItem('token');
  localStorage.removeItem('username');
}

/** Sets user preference for auto-syncing to {@link state} and `localStorage` */
export function setAutoSync(enabled: boolean): void {
  state.autoSyncEnabled = enabled;
  localStorage.setItem('auto-sync', enabled ? 'true' : 'false');
}

// Routes //

export async function login(): Promise<void> {
  state.isLoading = true;

  const res = await tauriFetch<Record<string, string | Note[]>>('/login', 'POST', {
    username: state.username,
    password: state.password,
  });

  if (res.ok) {
    resetError();
    tauriEmit('login');

    await invoke('sync_all_local_notes', { notes: res.data.notes }).catch(console.error);
    await getAllNotes().catch(console.error);

    state.token = res.data.token as string;
    state.password = '';

    localStorage.setItem('username', state.username);
    localStorage.setItem('token', state.token);
  }

  state.isLoading = false;

  if (!res.ok) {
    state.error = {
      type: ErrorType.Auth,
      message: parseErrorRes(res),
    };

    console.error(res.data);
  }
}

export async function signup(): Promise<void> {
  state.isLoading = true;

  const res = await tauriFetch<Record<string, string>>('/signup', 'POST', {
    username: state.username,
    password: state.password,
    notes: noteState.notes,
  });

  state.isLoading = false;

  if (res.ok) {
    resetError();
    tauriEmit('login');

    state.token = res.data.token;
    state.password = '';

    localStorage.setItem('username', state.username);
    localStorage.setItem('token', state.token);
  } else {
    state.error = {
      type: ErrorType.Auth,
      message: parseErrorRes(res),
    };

    console.error(res.data);
  }
}

export async function logout(): Promise<void> {
  if (!state.token) return; // Prevent bug where event.emit triggers event.listen

  state.isLoading = true;

  const res = await tauriFetch<Record<string, never | string>>('/logout', 'POST', {
    username: state.username,
    token: state.token,
  });

  state.isLoading = false;

  if (res.ok) {
    resetError();
    tauriEmit('logout');
    clientSideLogout();
  } else {
    state.error = {
      type: ErrorType.Logout,
      message: parseErrorRes(res),
    };
  }
}

export async function pull(): Promise<void> {
  state.isLoading = true;

  const res = await tauriFetch<Record<string, string | Note[]>>('/notes', 'POST', {
    username: state.username,
    token: state.token,
  });

  if (res.ok) {
    resetError();
    await invoke('sync_all_local_notes', { notes: res.data.notes }).catch(console.error);
    await getAllNotes().catch(console.error);
  }

  state.isLoading = false;

  if (!res.ok) {
    state.error = {
      type: ErrorType.Pull,
      message: parseErrorRes(res),
    };

    // User not found
    if (res.status === 404) clientSideLogout();

    console.error(res.data);
  }
}

export async function push(): Promise<void> {
  state.isLoading = true;

  const res = await tauriFetch<Record<string, never | string>>('/notes', 'PUT', {
    username: state.username,
    token: state.token,
    notes: noteState.notes,
  });

  state.isLoading = false;

  if (res.ok) {
    resetError();
    state.hasUnsyncedNotes = false;
  } else {
    state.error = {
      type: ErrorType.Push,
      message: parseErrorRes(res),
    };

    console.error(res.data);
  }
}

document.addEventListener('note-unsynced', () => {
  if (state.token) {
    state.hasUnsyncedNotes = true;
  }
});
