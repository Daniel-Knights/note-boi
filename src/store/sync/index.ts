import { http } from '@tauri-apps/api';
import type { Response } from '@tauri-apps/api/http';
import { reactive } from 'vue';

import { STORAGE_KEYS } from '../../constant';
import { isDev } from '../../utils';
import { UnsyncedNoteIds, unsyncedNoteIds } from './note';

export enum ErrorType {
  None,
  Auth,
  Push,
  Pull,
  Logout,
}

export const syncState = reactive({
  username: localStorage.getItem(STORAGE_KEYS.USERNAME) || '',
  password: '',
  token: localStorage.getItem(STORAGE_KEYS.TOKEN) || '',
  isLoading: false,
  isLogin: true, // For switching login/signup form
  error: {
    type: ErrorType.None,
    message: '',
  },
  unsyncedNoteIds: <UnsyncedNoteIds>{
    new: unsyncedNoteIds.new || '',
    edited: new Set<string>(unsyncedNoteIds.edited),
    deleted: new Set<string>(unsyncedNoteIds.deleted),
    get size() {
      return this.edited.size + this.deleted.size;
    },
    clear() {
      this.new = '';
      this.edited.clear();
      this.deleted.clear();
      localStorage.removeItem(STORAGE_KEYS.UNSYNCED);
    },
    add(ids): void {
      if (ids.new !== undefined) this.new = ids.new;

      ids.edited?.forEach((id) => this.edited.add(id));

      // Only track deleted notes if logged in
      ids.deleted?.forEach((id) => {
        this.deleted.add(id);

        if (this.edited.has(id)) {
          this.edited.delete(id);
        }
      });

      if (this.edited.has(this.new) || this.deleted.has(this.new)) {
        this.new = '';
      }

      localStorage.setItem(
        STORAGE_KEYS.UNSYNCED,
        JSON.stringify({
          new: this.new,
          edited: [...this.edited],
          deleted: [...this.deleted],
        })
      );
    },
  },
});

// Utils //

/** Parses an error response and returns a formatted message. */
export function parseErrorRes(res: Response<Record<string, unknown>>): string {
  const unknownErrorMessage = 'Unknown error, please try again';

  if (!res.data) return unknownErrorMessage;

  return typeof res.data.error === 'string' ? res.data.error : unknownErrorMessage;
}

/** Resets {@link state.error}. */
export function resetError(): void {
  syncState.error = { type: ErrorType.None, message: '' };
}

/** Wrapper for {@link http.fetch}. */
export function tauriFetch<T>(
  endpoint: string,
  method: 'POST' | 'PUT',
  payload?: Record<string, unknown>
): Promise<Response<T>> {
  const baseUrl = isDev()
    ? 'http://localhost:8000'
    : 'https://note-boi-server.herokuapp.com';

  return http.fetch<T>(`${baseUrl}/api${endpoint}`, {
    method,
    body: {
      type: 'Json',
      payload,
    },
  });
}

/** Catches hanging requests (e.g. due to server error). */
export function catchHang(err: unknown, type: ErrorType): void {
  console.error(err);

  syncState.isLoading = false;
  syncState.error = {
    type,
    message: 'Request failed',
  };
}

export { autoPush, push, pull, syncNotes, UnsyncedNoteIds } from './note';
export { clientSideLogout, login, signup, logout } from './auth';
export { deleteAccount } from './account';
