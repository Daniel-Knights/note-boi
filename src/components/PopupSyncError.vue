<template>
  <Popup @close="emit('close')">
    <div class="sync-error__message" data-test-id="error-message">
      Error: {{ state.error.message }}
    </div>
    <button @click="tryAgain" class="sync-error__button" data-test-id="try-again">
      Try again?
    </button>
  </Popup>
</template>

<script lang="ts" setup>
import { ErrorType, logout, pull, push, resetError, state } from '../store/sync';

import Popup from './Popup.vue';

const emit = defineEmits(['close']);

function tryAgain() {
  switch (state.error.type) {
    case ErrorType.Push:
      push();
      break;
    case ErrorType.Pull:
      pull();
      break;
    case ErrorType.Logout:
      logout();
      break;
    // no default
  }

  resetError();
  emit('close');
}
</script>

<style lang="scss" scoped>
.sync-error__message {
  font-size: 20px;
  font-weight: 600;
}

.sync-error__button {
  margin-top: 12px;
  text-decoration: underline;

  &:hover {
    cursor: pointer;
    text-decoration: none;
  }
}
</style>
