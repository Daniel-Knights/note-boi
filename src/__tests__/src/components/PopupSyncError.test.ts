import { mount } from '@vue/test-utils';
import { nextTick } from 'vue';

import * as s from '../../../store/sync';
import { AppError, ERROR_CODE } from '../../../classes';
import { mockApi } from '../../api';
import {
  assertAppError,
  findByTestId,
  getByTestId,
  resolveImmediate,
  waitUntil,
} from '../../utils';

import Popup from '../../../components/Popup.vue';
import PopupSyncError from '../../../components/PopupSyncError.vue';

function mountPopupSyncError() {
  return mount(PopupSyncError, {
    global: {
      stubs: { teleport: true },
    },
  });
}

describe('PopupSyncError', () => {
  it('Mounts', async () => {
    const { calls, promises } = mockApi();
    const wrapper = mountPopupSyncError();
    const errorMessage = 'I am a sync error';

    s.syncState.appError = new AppError({
      code: ERROR_CODE.SYNC,
      message: errorMessage,
    });

    await Promise.all(promises);

    assert.isTrue(wrapper.isVisible());
    assert.strictEqual(calls.size, 0);

    const errorMessageWrapper = getByTestId(wrapper, 'error-message');
    assert.strictEqual(errorMessageWrapper.text(), `Error: ${errorMessage}`);
  });

  it('Emits close', () => {
    const wrapper = mountPopupSyncError();
    wrapper.getComponent(Popup).vm.$emit('close');

    assert.lengthOf(wrapper.emitted('close')!, 1);
  });

  it('Retries', async () => {
    mockApi();

    const retryMock = vi.fn();
    const resetErrorSpy = vi.spyOn(s, 'resetAppError');

    s.syncState.appError = new AppError({
      code: ERROR_CODE.SYNC,
      retry: { fn: retryMock },
    });

    const wrapper = mountPopupSyncError();
    const tryAgainButton = getByTestId(wrapper, 'try-again');

    await tryAgainButton.trigger('click');
    await waitUntil(() => !s.syncState.loadingCount);

    expect(resetErrorSpy).toHaveBeenCalledOnce();
    expect(retryMock).toHaveBeenCalledOnce();

    assertAppError();
    assert.lengthOf(wrapper.emitted('close')!, 1);
  });

  it('No try again button when no retry function', () => {
    mockApi();

    s.syncState.appError = new AppError({
      code: ERROR_CODE.SYNC,
    });

    const wrapper = mountPopupSyncError();

    assert.isFalse(findByTestId(wrapper, 'try-again').exists());
  });

  it('Ignores', async () => {
    const { calls, promises } = mockApi();

    const wrapper = mountPopupSyncError();
    const ignoreButton = getByTestId(wrapper, 'ignore');
    const resetErrorSpy = vi.spyOn(s, 'resetAppError');
    const clientSideLogoutSpy = vi.spyOn(s, 'clientSideLogout');

    await ignoreButton.trigger('click');
    await nextTick();
    await resolveImmediate();
    await Promise.all(promises);

    expect(resetErrorSpy).toHaveBeenCalledOnce();
    expect(clientSideLogoutSpy).toHaveBeenCalledOnce();

    assertAppError();
    assert.lengthOf(wrapper.emitted('close')!, 1);
    assert.strictEqual(calls.size, 2);
    assert.isTrue(calls.invoke.has('delete_access_token'));
    assert.deepEqual(calls.invoke[0]!.calledWith, { username: '' });
    assert.isTrue(calls.emits.has('auth'));
  });

  it('Shows default error message when none provided', async () => {
    const { calls, promises } = mockApi();
    const wrapper = mountPopupSyncError();

    s.syncState.appError = new AppError({
      code: ERROR_CODE.SYNC,
    });

    await Promise.all(promises);

    assert.isTrue(wrapper.isVisible());
    assert.strictEqual(calls.size, 0);

    const errorMessageWrapper = getByTestId(wrapper, 'error-message');
    assert.strictEqual(errorMessageWrapper.text(), 'Error: Something went wrong');
  });
});
