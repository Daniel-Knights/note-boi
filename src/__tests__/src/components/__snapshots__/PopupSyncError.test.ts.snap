// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`PopupSyncError > Emits close > KeyStore.getKey() 1`] = `undefined`;

exports[`PopupSyncError > Emits close > Storage.get('MENU_WIDTH') 1`] = `null`;

exports[`PopupSyncError > Emits close > Storage.get('THEME') 1`] = `null`;

exports[`PopupSyncError > Emits close > Storage.get('UPDATE_SEEN') 1`] = `null`;

exports[`PopupSyncError > Emits close > Storage.get('UPDATE_STRATEGY') 1`] = `null`;

exports[`PopupSyncError > Emits close > Storage.get('USERNAME') 1`] = `null`;

exports[`PopupSyncError > Emits close > allCalls 1`] = `[]`;

exports[`PopupSyncError > Emits close > document.body.outerHTML 1`] = `"<body class="theme--system"></body>"`;

exports[`PopupSyncError > Emits close > mockDb 1`] = `
{
  "encryptedNotes": 10,
  "users": {
    "d": "1",
  },
}
`;

exports[`PopupSyncError > Emits close > noteState 1`] = `
{
  "extraSelectedNotes": [],
  "notes": [],
  "selectedNote": {
    "content": {
      "body": "",
      "delta": {
        "ops": [],
      },
      "title": "",
    },
    "id": "id",
    "timestamp": 0,
  },
}
`;

exports[`PopupSyncError > Emits close > openedPopup 1`] = `undefined`;

exports[`PopupSyncError > Emits close > selectedTheme 1`] = `"System"`;

exports[`PopupSyncError > Emits close > syncState 1`] = `
{
  "appError": AppError {
    "code": 0,
    "display": undefined,
    "message": undefined,
    "originalError": undefined,
    "retry": [Function],
    "retryConfig": undefined,
  },
  "isLoggedIn": false,
  "loadingCount": 0,
  "newPassword": "",
  "password": "",
  "unsyncedNoteIds": {
    "add": [Function],
    "clear": [Function],
    "deleted": Set {},
    "edited": Set {},
    "new": "",
    "size": 0,
  },
  "username": "",
}
`;

exports[`PopupSyncError > Emits close > updateState 1`] = `
{
  "isAvailable": false,
  "isDownloading": false,
  "strategy": "manual",
}
`;

exports[`PopupSyncError > Ignores > KeyStore.getKey() 1`] = `undefined`;

exports[`PopupSyncError > Ignores > Storage.get('MENU_WIDTH') 1`] = `null`;

exports[`PopupSyncError > Ignores > Storage.get('THEME') 1`] = `null`;

exports[`PopupSyncError > Ignores > Storage.get('UPDATE_SEEN') 1`] = `null`;

exports[`PopupSyncError > Ignores > Storage.get('UPDATE_STRATEGY') 1`] = `null`;

exports[`PopupSyncError > Ignores > Storage.get('USERNAME') 1`] = `null`;

exports[`PopupSyncError > Ignores > allCalls 1`] = `
[
  [
    "invoke",
    {
      "calledWith": {
        "username": "",
      },
      "name": "delete_access_token",
    },
  ],
  [
    "emits",
    {
      "calledWith": {
        "data": {
          "is_logged_in": false,
        },
        "isFrontendEmit": true,
      },
      "name": "auth",
    },
  ],
]
`;

exports[`PopupSyncError > Ignores > document.body.outerHTML 1`] = `"<body class="theme--system"></body>"`;

exports[`PopupSyncError > Ignores > mockDb 1`] = `
{
  "encryptedNotes": 10,
  "users": {
    "d": "1",
  },
}
`;

exports[`PopupSyncError > Ignores > noteState 1`] = `
{
  "extraSelectedNotes": [],
  "notes": [],
  "selectedNote": {
    "content": {
      "body": "",
      "delta": {
        "ops": [],
      },
      "title": "",
    },
    "id": "id",
    "timestamp": 0,
  },
}
`;

exports[`PopupSyncError > Ignores > openedPopup 1`] = `undefined`;

exports[`PopupSyncError > Ignores > selectedTheme 1`] = `"System"`;

exports[`PopupSyncError > Ignores > syncState 1`] = `
{
  "appError": AppError {
    "code": 0,
    "display": undefined,
    "message": undefined,
    "originalError": undefined,
    "retry": [Function],
    "retryConfig": undefined,
  },
  "isLoggedIn": false,
  "loadingCount": 0,
  "newPassword": "",
  "password": "",
  "unsyncedNoteIds": {
    "add": [Function],
    "clear": [Function],
    "deleted": Set {},
    "edited": Set {},
    "new": "",
    "size": 0,
  },
  "username": "",
}
`;

exports[`PopupSyncError > Ignores > updateState 1`] = `
{
  "isAvailable": false,
  "isDownloading": false,
  "strategy": "manual",
}
`;

exports[`PopupSyncError > Mounts > KeyStore.getKey() 1`] = `undefined`;

exports[`PopupSyncError > Mounts > Storage.get('MENU_WIDTH') 1`] = `null`;

exports[`PopupSyncError > Mounts > Storage.get('THEME') 1`] = `null`;

exports[`PopupSyncError > Mounts > Storage.get('UPDATE_SEEN') 1`] = `null`;

exports[`PopupSyncError > Mounts > Storage.get('UPDATE_STRATEGY') 1`] = `null`;

exports[`PopupSyncError > Mounts > Storage.get('USERNAME') 1`] = `null`;

exports[`PopupSyncError > Mounts > allCalls 1`] = `[]`;

exports[`PopupSyncError > Mounts > document.body.outerHTML 1`] = `"<body class="theme--system"></body>"`;

exports[`PopupSyncError > Mounts > mockDb 1`] = `
{
  "encryptedNotes": 10,
  "users": {
    "d": "1",
  },
}
`;

exports[`PopupSyncError > Mounts > noteState 1`] = `
{
  "extraSelectedNotes": [],
  "notes": [],
  "selectedNote": {
    "content": {
      "body": "",
      "delta": {
        "ops": [],
      },
      "title": "",
    },
    "id": "id",
    "timestamp": 0,
  },
}
`;

exports[`PopupSyncError > Mounts > openedPopup 1`] = `undefined`;

exports[`PopupSyncError > Mounts > selectedTheme 1`] = `"System"`;

exports[`PopupSyncError > Mounts > syncState 1`] = `
{
  "appError": AppError {
    "code": 6,
    "display": undefined,
    "message": "I am a sync error",
    "originalError": undefined,
    "retry": [Function],
    "retryConfig": undefined,
  },
  "isLoggedIn": false,
  "loadingCount": 0,
  "newPassword": "",
  "password": "",
  "unsyncedNoteIds": {
    "add": [Function],
    "clear": [Function],
    "deleted": Set {},
    "edited": Set {},
    "new": "",
    "size": 0,
  },
  "username": "",
}
`;

exports[`PopupSyncError > Mounts > updateState 1`] = `
{
  "isAvailable": false,
  "isDownloading": false,
  "strategy": "manual",
}
`;

exports[`PopupSyncError > No try again button when no retry function > KeyStore.getKey() 1`] = `undefined`;

exports[`PopupSyncError > No try again button when no retry function > Storage.get('MENU_WIDTH') 1`] = `null`;

exports[`PopupSyncError > No try again button when no retry function > Storage.get('THEME') 1`] = `null`;

exports[`PopupSyncError > No try again button when no retry function > Storage.get('UPDATE_SEEN') 1`] = `null`;

exports[`PopupSyncError > No try again button when no retry function > Storage.get('UPDATE_STRATEGY') 1`] = `null`;

exports[`PopupSyncError > No try again button when no retry function > Storage.get('USERNAME') 1`] = `null`;

exports[`PopupSyncError > No try again button when no retry function > allCalls 1`] = `[]`;

exports[`PopupSyncError > No try again button when no retry function > document.body.outerHTML 1`] = `"<body class="theme--system"></body>"`;

exports[`PopupSyncError > No try again button when no retry function > mockDb 1`] = `
{
  "encryptedNotes": 10,
  "users": {
    "d": "1",
  },
}
`;

exports[`PopupSyncError > No try again button when no retry function > noteState 1`] = `
{
  "extraSelectedNotes": [],
  "notes": [],
  "selectedNote": {
    "content": {
      "body": "",
      "delta": {
        "ops": [],
      },
      "title": "",
    },
    "id": "id",
    "timestamp": 0,
  },
}
`;

exports[`PopupSyncError > No try again button when no retry function > openedPopup 1`] = `undefined`;

exports[`PopupSyncError > No try again button when no retry function > selectedTheme 1`] = `"System"`;

exports[`PopupSyncError > No try again button when no retry function > syncState 1`] = `
{
  "appError": AppError {
    "code": 6,
    "display": undefined,
    "message": undefined,
    "originalError": undefined,
    "retry": [Function],
    "retryConfig": undefined,
  },
  "isLoggedIn": false,
  "loadingCount": 0,
  "newPassword": "",
  "password": "",
  "unsyncedNoteIds": {
    "add": [Function],
    "clear": [Function],
    "deleted": Set {},
    "edited": Set {},
    "new": "",
    "size": 0,
  },
  "username": "",
}
`;

exports[`PopupSyncError > No try again button when no retry function > updateState 1`] = `
{
  "isAvailable": false,
  "isDownloading": false,
  "strategy": "manual",
}
`;

exports[`PopupSyncError > Retries > KeyStore.getKey() 1`] = `undefined`;

exports[`PopupSyncError > Retries > Storage.get('MENU_WIDTH') 1`] = `null`;

exports[`PopupSyncError > Retries > Storage.get('THEME') 1`] = `null`;

exports[`PopupSyncError > Retries > Storage.get('UPDATE_SEEN') 1`] = `null`;

exports[`PopupSyncError > Retries > Storage.get('UPDATE_STRATEGY') 1`] = `null`;

exports[`PopupSyncError > Retries > Storage.get('USERNAME') 1`] = `null`;

exports[`PopupSyncError > Retries > allCalls 1`] = `[]`;

exports[`PopupSyncError > Retries > document.body.outerHTML 1`] = `"<body class="theme--system"></body>"`;

exports[`PopupSyncError > Retries > mockDb 1`] = `
{
  "encryptedNotes": 10,
  "users": {
    "d": "1",
  },
}
`;

exports[`PopupSyncError > Retries > noteState 1`] = `
{
  "extraSelectedNotes": [],
  "notes": [],
  "selectedNote": {
    "content": {
      "body": "",
      "delta": {
        "ops": [],
      },
      "title": "",
    },
    "id": "id",
    "timestamp": 0,
  },
}
`;

exports[`PopupSyncError > Retries > openedPopup 1`] = `undefined`;

exports[`PopupSyncError > Retries > selectedTheme 1`] = `"System"`;

exports[`PopupSyncError > Retries > syncState 1`] = `
{
  "appError": AppError {
    "code": 0,
    "display": undefined,
    "message": undefined,
    "originalError": undefined,
    "retry": [Function],
    "retryConfig": undefined,
  },
  "isLoggedIn": false,
  "loadingCount": 0,
  "newPassword": "",
  "password": "",
  "unsyncedNoteIds": {
    "add": [Function],
    "clear": [Function],
    "deleted": Set {},
    "edited": Set {},
    "new": "",
    "size": 0,
  },
  "username": "",
}
`;

exports[`PopupSyncError > Retries > updateState 1`] = `
{
  "isAvailable": false,
  "isDownloading": false,
  "strategy": "manual",
}
`;

exports[`PopupSyncError > Shows default error message when none provided > KeyStore.getKey() 1`] = `undefined`;

exports[`PopupSyncError > Shows default error message when none provided > Storage.get('MENU_WIDTH') 1`] = `null`;

exports[`PopupSyncError > Shows default error message when none provided > Storage.get('THEME') 1`] = `null`;

exports[`PopupSyncError > Shows default error message when none provided > Storage.get('UPDATE_SEEN') 1`] = `null`;

exports[`PopupSyncError > Shows default error message when none provided > Storage.get('UPDATE_STRATEGY') 1`] = `null`;

exports[`PopupSyncError > Shows default error message when none provided > Storage.get('USERNAME') 1`] = `null`;

exports[`PopupSyncError > Shows default error message when none provided > allCalls 1`] = `[]`;

exports[`PopupSyncError > Shows default error message when none provided > document.body.outerHTML 1`] = `"<body class="theme--system"></body>"`;

exports[`PopupSyncError > Shows default error message when none provided > mockDb 1`] = `
{
  "encryptedNotes": 10,
  "users": {
    "d": "1",
  },
}
`;

exports[`PopupSyncError > Shows default error message when none provided > noteState 1`] = `
{
  "extraSelectedNotes": [],
  "notes": [],
  "selectedNote": {
    "content": {
      "body": "",
      "delta": {
        "ops": [],
      },
      "title": "",
    },
    "id": "id",
    "timestamp": 0,
  },
}
`;

exports[`PopupSyncError > Shows default error message when none provided > openedPopup 1`] = `undefined`;

exports[`PopupSyncError > Shows default error message when none provided > selectedTheme 1`] = `"System"`;

exports[`PopupSyncError > Shows default error message when none provided > syncState 1`] = `
{
  "appError": AppError {
    "code": 6,
    "display": undefined,
    "message": undefined,
    "originalError": undefined,
    "retry": [Function],
    "retryConfig": undefined,
  },
  "isLoggedIn": false,
  "loadingCount": 0,
  "newPassword": "",
  "password": "",
  "unsyncedNoteIds": {
    "add": [Function],
    "clear": [Function],
    "deleted": Set {},
    "edited": Set {},
    "new": "",
    "size": 0,
  },
  "username": "",
}
`;

exports[`PopupSyncError > Shows default error message when none provided > updateState 1`] = `
{
  "isAvailable": false,
  "isDownloading": false,
  "strategy": "manual",
}
`;
